### ISSUE TITLE
Since Update to Pi-hole v5.2.4 Web Interface v5.3.1 FTL v5.5.1 I have a lot of Unknown (0) entries in the Query Log

### ISSUE THREAD

**Issue by MeshuggahMischell** [pi-hole/web#1713]
### Versions

- Pi-hole:  Pi-hole version is v5.2.4 (Latest: v5.2.4)
- AdminLTE:  AdminLTE version is v5.3.1 (Latest: v5.3.1)
- FTL:  FTL version is v5.5.1 (Latest: v5.5.1)

### Platform

- OS and version:   Linux 5.4.79-v7+ armv7l

- Platform:  RaspberryPi / DietPi

## Debug Token

- URL:  https://tricorder.pi-hole.net/d9o5hp5uoo

### Screenshots

![pihole](https://user-images.githubusercontent.com/51913770/105335289-2b276e00-5bd8-11eb-8d57-c137eb85e183.jpg)

### Additional context

Since Update to Pi-hole v5.2.4 Web Interface v5.3.1 FTL v5.5.1 I have a lot of Unknown (0) entries in the Query Log

I never seen entries like that before, is this maybe an bug or error in the new Version ?

**Comment by DL6ER** [comment]
@MeshuggahMischell @r2ixuz Could you try to bring these queries up on the long-term query log and check if they have an unknown status there as well?

Check out http://pi.hole/admin/db_queries.php and try to narrow down the time interval enough so you can see them there. How does the status column of these queries look like there?

**Comment by MeshuggahMischell** [comment]
In the longterm query log they are shown as OK 
I drilled down to the same time as in the first Screenshot.
![PiHoleLongtimequery](https://user-images.githubusercontent.com/51913770/105378775-70fe2980-5c0c-11eb-9775-f342a3cf0934.png)

**Comment by Daxtorim** [comment]
They don't appear at all in the long-term data. Notice the timestamps:
Query Log:
![grafik](https://user-images.githubusercontent.com/72616153/105379144-d6521a80-5c0c-11eb-8dba-09370546e479.png)

Long-term data:
![grafik](https://user-images.githubusercontent.com/72616153/105379161-d9e5a180-5c0c-11eb-9f4f-ea781c73e473.png)

Addition: All queries with status _Unknown (0)_ are for CNAMEs.

**Comment by Daxtorim** [comment]
I see a lot more queries with status _Unknown (0)_ after restarting pihole-FTL. The screenshots I posted were after pihole-FTL ran for over a day and for those queries my addition above was true, they were all CNAMEs. They were also **not** repeated afterwards.

But now after restarting pihole-FTL, there are queries with status _Unknown (0)_ that simply weren't shown before. Those "new" queries are also no longer only CNAMEs. They are however only for queries that were originally forwarded, even when they were blocked after CNAME inspection.
![grafik](https://user-images.githubusercontent.com/72616153/105384976-1f0cd200-5c13-11eb-8870-137ef5e1d256.png)
I have not found any query with unknown status that was originally blocked by gravity or regex.

**Comment by DL6ER** [comment]
There is some bug in the upstream `dnsmasq` which will likely trigger a new release sooner or later. This *may* be causing this. As we didn't touch anything in the code related to this, this is actually pretty reasonable. We track the updates here: [pi-hole/FTL#update/dnsmasq-v2.84](https://github.com/pi-hole/FTL/tree/update/dnsmasq-v2.84)

Could you meanwhile check your `/var/log/pihole.log` for the lines related to these unknown queries?

If nothing obvious comes up, please also add
```
DEBUG_QUERIES=true
DEBUG_FLAGS=true
DEBUG_DNSMASQ_LINES=true
```
to the file `/etc/pihole/pihole-FTL.conf` (create if it does not exist) and run
```
pihole restartdns
```

Next time you see `Unknown (0)` queries, please check again `/var/log/pihole.log` for the related lines. In addition, please check `/var/log/pihole-FTL.log` for these lines. Queries show up like
```
[2021-01-23 08:48:42.614 1206376M] **** new UDP query[A] query "api.github.com" from enp2s0:192.168.2.224 (ID 555, FTL 19389, /home/me/FTL/src/dnsmasq/forward.c:1623)
```
The important bit to memorize here is `ID 555`. Please run (**substitute the ID you found below**)
```
grep -A1 "ID 555" /var/log/pihole.log
```
and share the output.

Make sure to remove the debug lines again if you are short on available disk space (this can make the log grow real quick!).

**Comment by DL6ER** [comment]
There is a bug in `dnsmasq` upstream causing some queries to just fail. This *may* be causing this. As we didn't touch anything in the code related to this, this is actually pretty reasonable. We track the updates here: [pi-hole/FTL#update/dnsmasq-v2.84](https://github.com/pi-hole/FTL/tree/update/dnsmasq-v2.84)

Meanwhile, please your `/var/log/pihole.log` for the queries you see as `Unknown (0)` here (plus some lines afterwards). How do they look like?

**Comment by libertyx82** [comment]
Here the Lines in the pihole.log related to the query of alive.github.com

```
Jan 23 18:30:15 dnsmasq[724]: query[A] alive.github.com from fd00::xxxx:yyy:aaaa:bbbb
Jan 23 18:30:15 dnsmasq[724]: forwarded alive.github.com to 127.0.0.1
Jan 23 18:30:15 dnsmasq[724]: query[AAAA] alive.github.com from fd00::xxxx:yyy:aaaa:bbbb
Jan 23 18:30:15 dnsmasq[724]: forwarded alive.github.com to 127.0.0.1
Jan 23 18:30:15 dnsmasq[724]: query[AAAA] alive.github.com from 192.168.2.10
Jan 23 18:30:15 dnsmasq[724]: query[A] alive.github.com from 192.168.2.10
Jan 23 18:30:16 dnsmasq[724]: reply alive.github.com is <CNAME>
Jan 23 18:30:16 dnsmasq[724]: reply live.github.com is 140.82.112.25
Jan 23 18:30:16 dnsmasq[724]: failed to send packet: Network is unreachable
Jan 23 18:30:16 dnsmasq[724]: reply alive.github.com is <CNAME>
Jan 23 18:30:16 dnsmasq[724]: reply live.github.com is NODATA-IPv6
Jan 23 18:30:16 dnsmasq[724]: failed to send packet: Network is unreachable
```
![image](https://user-images.githubusercontent.com/22996350/105609479-eb17e500-5da9-11eb-864a-cf38c1cd3b65.png)


```
[2021-01-23 18:40:42.318 2110M] **** new UDP query[AAAA] query "github.githubassets.com" from ens192:fd00::xxxx:yyy:aaaa:bbbb (ID 64, FTL 45873, /root/project/src/dnsmasq/forward.c:1631)
[2021-01-23 18:40:42.318 2110M] github.githubassets.com is not known
[2021-01-23 18:40:42.318 2110M] **** forwarded github.githubassets.com to ::1#5153 (ID 64, /root/project/src/dnsmasq/forward.c:605)
[2021-01-23 18:40:42.360 2110M] **** new UDP query[AAAA] query "github.githubassets.com" from ens192:192.168.2.10 (ID 66, FTL 45875, /root/project/src/dnsmasq/forward.c:1624)
[2021-01-23 18:40:42.360 2110M] github.githubassets.com is not known
[2021-01-23 18:40:42.411 2110M] **** got reply github.githubassets.com is (NODATA) (ID 64, /root/project/src/dnsmasq/cache.c:494)
```

```
Jan 23 18:40:42 dnsmasq[2110]: query[AAAA] github.githubassets.com from fd00::xxxx:yyy:aaaa:bbbb (/root/project/src/dnsmasq/forward.c:1629)
Jan 23 18:40:42 dnsmasq[2110]: forwarded github.githubassets.com to ::1 (/root/project/src/dnsmasq/forward.c:603)
Jan 23 18:40:42 dnsmasq[2110]: query[AAAA] github.githubassets.com from 192.168.2.10 (/root/project/src/dnsmasq/forward.c:1622)
Jan 23 18:40:42 dnsmasq[2110]: reply github.githubassets.com is NODATA-IPv6 (/root/project/src/dnsmasq/cache.c:493)
```
```
:/$ grep -A1 "ID 64" /var/log/pihole.log
:/$

:/$ grep -A1 "ID 66" /var/log/pihole.log
:/$

```

**Comment by Daxtorim** [comment]
![grafik](https://user-images.githubusercontent.com/72616153/105579293-f35e2900-5d85-11eb-876b-8b361b625df8.png)

---

/var/log/pihole.log :
```
Jan 23 14:12:46 dnsmasq[9472]: query[A] steamcommunity.com from fddc:704f:99b9:178:1c4:74c3:e0fb:ddb (/root/project/src/dnsmasq/forward.c:1629)
Jan 23 14:12:46 dnsmasq[9472]: forwarded steamcommunity.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:596)
Jan 23 14:12:46 dnsmasq[9472]: query[AAAA] steamcommunity.com from fddc:704f:99b9:178:1c4:74c3:e0fb:ddb (/root/project/src/dnsmasq/forward.c:1629)
Jan 23 14:12:46 dnsmasq[9472]: forwarded steamcommunity.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:596)
Jan 23 14:12:46 dnsmasq[9472]: query[AAAA] steamcommunity.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1622)
Jan 23 14:12:46 dnsmasq[9472]: query[A] steamcommunity.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1622)
Jan 23 14:12:46 dnsmasq[9472]: dnssec-query[DS] steamcommunity.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:1193)
Jan 23 14:12:46 dnsmasq[9472]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1284)
Jan 23 14:12:46 dnsmasq[9472]: reply steamcommunity.com is NODATA-IPv6 (/root/project/src/dnsmasq/cache.c:493)
Jan 23 14:12:46 dnsmasq[9472]: failed to send packet: Network is unreachable
Jan 23 14:12:47 dnsmasq[9472]: reply steamcommunity.com is no DS (/root/project/src/dnsmasq/dnssec.c:1053)
Jan 23 14:12:47 dnsmasq[9472]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1284)
Jan 23 14:12:47 dnsmasq[9472]: reply steamcommunity.com is 23.14.245.221 (/root/project/src/dnsmasq/cache.c:493)
Jan 23 14:12:47 dnsmasq[9472]: failed to send packet: Network is unreachable
```
The obvious anomaly here are the two instances of  `dnsmasq[9472]: failed to send packet: Network is unreachable`. My Pi is not reporting any issues regarding the network though, there is nothing in /var/log/syslog for that time period.  

---

/var/log/pihole-FTL.log :
```
╭[pi@pihole-unbound]-[~]
╰─╸$ grep -A1 "ID 327" /var/log/pihole-FTL.log
[2021-01-23 14:12:46.290 9472M] **** new UDP query[A] query "steamcommunity.com" from wlan0:fddc:704f:99b9:178:1c4:74c3:e0fb:ddb (ID 327, FTL 96788, /root/project/src/dnsmasq/forward.c:1631)
[2021-01-23 14:12:46.291 9472M] steamcommunity.com is not known
[2021-01-23 14:12:46.296 9472M] **** forwarded steamcommunity.com to 127.0.0.1#5335 (ID 327, /root/project/src/dnsmasq/forward.c:598)
[2021-01-23 14:12:46.297 9472M] **** new UDP query[AAAA] query "steamcommunity.com" from wlan0:fddc:704f:99b9:178:1c4:74c3:e0fb:ddb (ID 328, FTL 96789, /root/project/src/dnsmasq/forward.c:1631)
--
[2021-01-23 14:12:47.259 9472M] **** got DNSSEC details for steamcommunity.com: 2 (ID 327, /root/project/src/dnsmasq/forward.c:1285)
[2021-01-23 14:12:47.260 9472M] steamcommunity.com is known as not to be blocked
--
[2021-01-23 14:12:47.261 9472M] **** got reply steamcommunity.com is 23.14.245.221 (ID 327, /root/project/src/dnsmasq/cache.c:494)
[2021-01-23 14:12:47.261 9472M]      Flags: F_FORWARD F_IPV4

╭[pi@pihole-unbound]-[~]
╰─╸$ grep -A1 "ID 328" /var/log/pihole-FTL.log
[2021-01-23 14:12:46.297 9472M] **** new UDP query[AAAA] query "steamcommunity.com" from wlan0:fddc:704f:99b9:178:1c4:74c3:e0fb:ddb (ID 328, FTL 96789, /root/project/src/dnsmasq/forward.c:1631)
[2021-01-23 14:12:46.298 9472M] steamcommunity.com is not known
[2021-01-23 14:12:46.303 9472M] **** forwarded steamcommunity.com to 127.0.0.1#5335 (ID 328, /root/project/src/dnsmasq/forward.c:598)
[2021-01-23 14:12:46.542 9472M] **** new UDP query[AAAA] query "steamcommunity.com" from wlan0:192.168.178.20 (ID 329, FTL 96790, /root/project/src/dnsmasq/forward.c:1624)
--
[2021-01-23 14:12:46.844 9472M] **** got DNSSEC details for steamcommunity.com: 2 (ID 328, /root/project/src/dnsmasq/forward.c:1285)
[2021-01-23 14:12:46.844 9472M] **** got reply steamcommunity.com is (NODATA) (ID 328, /root/project/src/dnsmasq/cache.c:494)
[2021-01-23 14:12:46.845 9472M]      Flags: F_FORWARD F_NEG F_IPV6

╭[pi@pihole-unbound]-[~]
╰─╸$ grep -A1 "ID 329" /var/log/pihole-FTL.log
[2021-01-23 14:12:46.542 9472M] **** new UDP query[AAAA] query "steamcommunity.com" from wlan0:192.168.178.20 (ID 329, FTL 96790, /root/project/src/dnsmasq/forward.c:1624)
[2021-01-23 14:12:46.543 9472M] steamcommunity.com is not known

╭[pi@pihole-unbound]-[~]
╰─╸$ grep -A1 "ID 330" /var/log/pihole-FTL.log
[2021-01-23 14:12:46.545 9472M] **** new UDP query[A] query "steamcommunity.com" from wlan0:192.168.178.20 (ID 330, FTL 96791, /root/project/src/dnsmasq/forward.c:1624)
[2021-01-23 14:12:46.547 9472M] steamcommunity.com is not known
```

**Comment by DL6ER** [comment]
> `failed to send packet: Network is unreachable`

Yeah, this is exactly the `dnsmasq` bug reported on the mailing list (yesterday?).

@libertyx82 @Daxtorim  Please try
```
pihole checkout ftl update/dnsmasq-v2.84
```
and see if the fix works for you as well (absence of `Network is unreachable` and `Unknown (0)` queries).

According to Simon Kelley, it looks like if two queries are combined (because they ask the same question) then dnsmasq can get confused when it comes to return the answer, the reply to the second query can be sent via the socket that the first one arrived on. That's normally OK, but if the first query arrives via IPv4 and the second via IPv6, for instance, then the bug is triggered.

**Comment by libertyx82** [comment]
done `FTL version is update/dnsmasq-v2.84 vDev-6fe533e (Latest: v5.5.1)`
but it still exists

![image](https://user-images.githubusercontent.com/22996350/105614298-d0a13400-5dc8-11eb-8b61-d52552904564.png)

```
[2021-01-23 22:11:39.799 1131M] **** new UDP query[A] query "clients4.google.com" from ens192:fd00::aaaa:bbbb:cccc:dddd (ID 486, FTL 47549, /root/project/src/dnsmasq/forward.c:1632)
[2021-01-23 22:11:39.799 1131M] clients4.google.com is not known
[2021-01-23 22:11:39.800 1131M] **** forwarded clients4.google.com to ::1#5153 (ID 486, /root/project/src/dnsmasq/forward.c:606)
[2021-01-23 22:11:39.800 1131M] **** new UDP query[AAAA] query "clients4.google.com" from ens192:fd00::aaaa:bbbb:cccc:dddd (ID 487, FTL 47550, /root/project/src/dnsmasq/forward.c:1632)
[2021-01-23 22:11:39.800 1131M] clients4.google.com is not known
[2021-01-23 22:11:39.800 1131M] **** forwarded clients4.google.com to ::1#5153 (ID 487, /root/project/src/dnsmasq/forward.c:606)
[2021-01-23 22:11:39.831 1131M] **** new UDP query[AAAA] query "clients4.google.com" from ens192:192.168.2.10 (ID 488, FTL 47551, /root/project/src/dnsmasq/forward.c:1625)
[2021-01-23 22:11:39.831 1131M] clients4.google.com is not known
[2021-01-23 22:11:39.831 1131M] **** new UDP query[A] query "clients4.google.com" from ens192:192.168.2.10 (ID 489, FTL 47552, /root/project/src/dnsmasq/forward.c:1625)
[2021-01-23 22:11:39.831 1131M] clients4.google.com is not known
[2021-01-23 22:11:39.887 1131M] clients4.google.com is known as not to be blocked (whitelisted)
[2021-01-23 22:11:39.887 1131M] CNAME clients4.google.com
[2021-01-23 22:11:39.887 1131M] **** got reply clients4.google.com is (CNAME) (ID 487, /root/project/src/dnsmasq/cache.c:494)
[2021-01-23 22:11:39.887 1131M] CNAME clients4.google.com ---> clients.l.google.com
[2021-01-23 22:11:39.897 1131M] clients4.google.com is known as not to be blocked (whitelisted)
[2021-01-23 22:11:39.897 1131M] CNAME clients4.google.com
[2021-01-23 22:11:39.897 1131M] **** got reply clients4.google.com is (CNAME) (ID 486, /root/project/src/dnsmasq/cache.c:494)
[2021-01-23 22:11:39.897 1131M] CNAME clients4.google.com ---> clients.l.google.com
```

```
Jan 23 22:11:39 dnsmasq[1131]: query[A] clients4.google.com from fd00::aaaa:bbbb:cccc:dddd
(/root/project/src/dnsmasq/forward.c:1630)
Jan 23 22:11:39 dnsmasq[1131]: forwarded clients4.google.com to ::1 (/root/project/src/dnsmasq/forward.c:604)
Jan 23 22:11:39 dnsmasq[1131]: query[AAAA] clients4.google.com from fd00::aaaa:bbbb:cccc:dddd (/root/project/src/dnsmasq/forward.c:1630)
Jan 23 22:11:39 dnsmasq[1131]: forwarded clients4.google.com to ::1 (/root/project/src/dnsmasq/forward.c:604)
Jan 23 22:11:39 dnsmasq[1131]: query[AAAA] clients4.google.com from 192.168.2.10 (/root/project/src/dnsmasq/forward.c:1623)
Jan 23 22:11:39 dnsmasq[1131]: query[A] clients4.google.com from 192.168.2.10 (/root/project/src/dnsmasq/forward.c:1623)
Jan 23 22:11:39 dnsmasq[1131]: reply clients4.google.com is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
Jan 23 22:11:39 dnsmasq[1131]: reply clients4.google.com is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
```

**Comment by Daxtorim** [comment]
Same as @libertyx82. Since switching branches I'm no longer seeing `failed to send packet: Network is unreachable` in /var/log/pihole.log but I'm still seeing queries with status *Unknown (0)* in the Query Log.

>if the first query arrives via IPv4 and the second via IPv6, for instance, then the bug is triggered

That seems to work now as intended, but pihole is only logging the one answer. From just looking at the pihole.log files you would think the second query was never even acknowledged and the Query Log certainly reflects that.

---

![grafik](https://user-images.githubusercontent.com/72616153/105615591-24644b00-5dd2-11eb-8fa0-2941f688cdca.png)

Four queries (2x`A`, 2x`AAAA`), but only two answers:
```
Jan 23 23:17:57 dnsmasq[26004]: query[A] officehomeblobs.blob.core.windows.net from fddc:704f:99b9:178:b587:20e5:57b4:130 (/root/project/src/dnsmasq/forward.c:1630)
Jan 23 23:17:57 dnsmasq[26004]: forwarded officehomeblobs.blob.core.windows.net to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:597)
Jan 23 23:17:57 dnsmasq[26004]: query[AAAA] officehomeblobs.blob.core.windows.net from fddc:704f:99b9:178:b587:20e5:57b4:130 (/root/project/src/dnsmasq/forward.c:1630)
Jan 23 23:17:57 dnsmasq[26004]: forwarded officehomeblobs.blob.core.windows.net to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:597)
Jan 23 23:17:57 dnsmasq[26004]: query[AAAA] officehomeblobs.blob.core.windows.net from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1623)
Jan 23 23:17:57 dnsmasq[26004]: query[A] officehomeblobs.blob.core.windows.net from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1623)
Jan 23 23:17:57 dnsmasq[26004]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1285)
Jan 23 23:17:57 dnsmasq[26004]: reply officehomeblobs.blob.core.windows.net is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
Jan 23 23:17:57 dnsmasq[26004]: reply blob.ch1prdstr06a.store.core.windows.net is 40.116.120.16 (/root/project/src/dnsmasq/cache.c:493)
Jan 23 23:17:57 dnsmasq[26004]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1285)
Jan 23 23:17:57 dnsmasq[26004]: reply officehomeblobs.blob.core.windows.net is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
Jan 23 23:17:57 dnsmasq[26004]: reply blob.ch1prdstr06a.store.core.windows.net is NODATA-IPv6 (/root/project/src/dnsmasq/cache.c:493)
```

**Comment by DL6ER** [comment]
> That seems to work now as intended, but pihole is only logging the one answer. From just looking at the pihole.log files you would think the second query was never even acknowledged and the Query Log certainly reflects that.

Yes, that's exactly the "issue". In fact, there is no bug anymore because `dnsmasq` chose to never even answer the first query. This is new behavior.

What is the expected behavior? ~~I guess the query should be green, it should be recorded that the query was forwarded but the reply column should stay as `N/A`~~

**edit** The `dnsmasq` log is not in agreement with what the code does. The queries *are* replied to eventually, however, this is not logged. I'm still thinking what would be the best solution for us.

**Comment by DL6ER** [comment]
@libertyx82 @Daxtorim  Okay, please update your `FTL` (on the same branch) and also run
```
pihole checkout web update/dnsmasq-v2.84
```

There should now be a new reply type (internal ID is 14), saying that this query was already forwarded (because one and the same query will never be forwarded twice with `dnsmasq` v2.83+). Once the reply is ready, FTL will update all the duplicated queries so the Query Log should now be complete, again. `dnsmasq` handles these "duplicated" queries somewhat sloppy, hence also the missing logging for the queries in `pihole.log`. FTL should, however, be able to handle this just fine now.

Example of what you should see (one of them showing as forwarded, the others showing as queries waiting for the reply):

![Screenshot at 2021-01-24 10-49-25](https://user-images.githubusercontent.com/16748619/105626715-e0089780-5e31-11eb-95c1-f82b13726abb.png)

----

The "already forwarded" queries are replied to instantly (0.0msec) once the reply from upstream is available. It would theoretically be possible to measure their timing as well, however, this would mean we'd have to add a separate timer to the queries. This would come at the costs of 8 bytes of extra memory per query (so roughly 200 KB of RAM on a typical Pi-hole setup) but it's not really worth it as it cannot be used as a real measure for anything: The original forwarding will already be timed and the timing of the duplicates would solely and somewhat arbitrarily depend on when the clients send their requests.

**Comment by libertyx82** [comment]
Did I do something wrong?

![image](https://user-images.githubusercontent.com/22996350/105627067-b1d88700-5e34-11eb-8d82-33eb5a8433fe.png)

```
:~$ pihole checkout ftl update/dnsmasq-v2.84
  Please note that changing branches severely alters your Pi-hole subsystems
  Features that work on the master branch, may not on a development branch
  This feature is NOT supported unless a Pi-hole developer explicitly asks!
  Have you read and understood this? [y/N] y

  [✓] Branch update/dnsmasq-v2.84 exists
  [✓] Downloading and Installing FTL
  [✓] Restarting pihole-FTL service...
  [✓] Enabling pihole-FTL service to start on reboot...
:~$ pihole checkout web update/dnsmasq-v2.84
  Please note that changing branches severely alters your Pi-hole subsystems
  Features that work on the master branch, may not on a development branch
  This feature is NOT supported unless a Pi-hole developer explicitly asks!
  Have you read and understood this? [y/N] y

  [✓] Fetching branches from https://github.com/pi-hole/AdminLTE.git
  [i] 23 branches available for Web Admin

  [✓] Switching to branch: 'update/dnsmasq-v2.84' from 'refs/heads/master'
Bereits aktuell.
:~$
```

**Comment by DL6ER** [comment]
@libertyx82 The `Unknown (14)` is showing that it works correctly on the backend side. Maybe you have to force-refresh the browser? Try `Strg + F5` (or `Ctrl + F5` depending on your keyboard label ;-) ).

**Comment by libertyx82** [comment]
After i wrote the comment i had the same idea and yes a force refresh does it :)

**Comment by Daxtorim** [comment]
Hmm:
![grafik](https://user-images.githubusercontent.com/72616153/105632334-cb88c700-5e52-11eb-8ed7-8843f6312265.png)
Did it actually avoid the block or is that a visual bug? Unfortunately I had logging disabled again.

I enabled logging and added the debug flags again. I'll report back when it comes up again.

**Comment by Daxtorim** [comment]
![grafik](https://user-images.githubusercontent.com/72616153/105633292-0c370f00-5e58-11eb-843b-896c58293811.png)

---

#### /var/log/pihole.log :
```
Jan 24 15:10:19 dnsmasq[10382]: query[A] settings-win.data.microsoft.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 15:10:19 dnsmasq[10382]: forwarded settings-win.data.microsoft.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:600)
Jan 24 15:10:19 dnsmasq[10382]: query[AAAA] settings-win.data.microsoft.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 15:10:19 dnsmasq[10382]: forwarded settings-win.data.microsoft.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:600)
Jan 24 15:10:19 dnsmasq[10382]: query[A] settings-win.data.microsoft.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1631)
Jan 24 15:10:19 dnsmasq[10382]: query[AAAA] settings-win.data.microsoft.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1631)
Jan 24 15:10:19 dnsmasq[10382]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1288)
Jan 24 15:10:19 dnsmasq[10382]: reply settings-win.data.microsoft.com is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
Jan 24 15:10:19 dnsmasq[10382]: reply settingsfd-geo.trafficmanager.net is blocked during CNAME inspection (/root/project/src/dnsmasq/rfc1035.c:742)
Jan 24 15:10:20 dnsmasq[10382]: validation result is INSECURE (/root/project/src/dnsmasq/forward.c:1288)
Jan 24 15:10:20 dnsmasq[10382]: reply settings-win.data.microsoft.com is <CNAME> (/root/project/src/dnsmasq/cache.c:493)
Jan 24 15:10:20 dnsmasq[10382]: reply settingsfd-geo.trafficmanager.net is blocked during CNAME inspection (/root/project/src/dnsmasq/rfc1035.c:742)
```

---

#### /var/log/pihole-FTL.log :
```
[2021-01-24 15:10:19.839 10382M] **** new UDP query[A] query "settings-win.data.microsoft.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 689, FTL 50023, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 15:10:19.839 10382M] settings-win.data.microsoft.com is known as not to be blocked
[2021-01-24 15:10:19.842 10382M] **** forwarded settings-win.data.microsoft.com to 127.0.0.1#5335 (ID 689, /root/project/src/dnsmasq/forward.c:602)
[2021-01-24 15:10:19.843 10382M] **** new UDP query[AAAA] query "settings-win.data.microsoft.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 690, FTL 50024, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 15:10:19.844 10382M] settings-win.data.microsoft.com is known as not to be blocked
[2021-01-24 15:10:19.846 10382M] **** forwarded settings-win.data.microsoft.com to 127.0.0.1#5335 (ID 690, /root/project/src/dnsmasq/forward.c:602)
[2021-01-24 15:10:19.867 10382M] **** new UDP query[A] query "settings-win.data.microsoft.com" from wlan0:192.168.178.20 (ID 691, FTL 50025, /root/project/src/dnsmasq/forward.c:1633)
[2021-01-24 15:10:19.867 10382M] settings-win.data.microsoft.com is not known
[2021-01-24 15:10:19.869 10382M] **** query for settings-win.data.microsoft.com is already in progress (ID 691)
[2021-01-24 15:10:19.870 10382M] **** new UDP query[AAAA] query "settings-win.data.microsoft.com" from wlan0:192.168.178.20 (ID 692, FTL 50026, /root/project/src/dnsmasq/forward.c:1633)
[2021-01-24 15:10:19.870 10382M] settings-win.data.microsoft.com is not known
[2021-01-24 15:10:19.876 10382M] **** query for settings-win.data.microsoft.com is already in progress (ID 692)
[2021-01-24 15:10:19.877 10382M] **** got DNSSEC details for settings-win.data.microsoft.com: 2 (ID 689, /root/project/src/dnsmasq/forward.c:1289)
[2021-01-24 15:10:19.878 10382M] settings-win.data.microsoft.com is known as not to be blocked
[2021-01-24 15:10:19.878 10382M] CNAME settings-win.data.microsoft.com
[2021-01-24 15:10:19.878 10382M] **** got reply settings-win.data.microsoft.com is (CNAME) (ID 689, /root/project/src/dnsmasq/cache.c:494)
[2021-01-24 15:10:19.879 10382M]      Flags: F_FORWARD F_CNAME
[2021-01-24 15:10:19.879 10382M] settingsfd-geo.trafficmanager.net is known as gravity blocked
[2021-01-24 15:10:19.879 10382M] CNAME settings-win.data.microsoft.com ---> settingsfd-geo.trafficmanager.net
[2021-01-24 15:10:19.880 10382M] CNAME header: Question was <IN> <A> settings-win.data.microsoft.com
[2021-01-24 15:10:19.881 10382M] **** query 50025 is duplicate of 50023
[2021-01-24 15:10:20.011 10382M] **** got DNSSEC details for settings-win.data.microsoft.com: 2 (ID 690, /root/project/src/dnsmasq/forward.c:1289)
[2021-01-24 15:10:20.012 10382M] settings-win.data.microsoft.com is known as not to be blocked
[2021-01-24 15:10:20.012 10382M] CNAME settings-win.data.microsoft.com
[2021-01-24 15:10:20.012 10382M] **** got reply settings-win.data.microsoft.com is (CNAME) (ID 690, /root/project/src/dnsmasq/cache.c:494)
[2021-01-24 15:10:20.013 10382M]      Flags: F_FORWARD F_CNAME
[2021-01-24 15:10:20.013 10382M] settingsfd-geo.trafficmanager.net is known as gravity blocked
[2021-01-24 15:10:20.013 10382M] CNAME settings-win.data.microsoft.com ---> settingsfd-geo.trafficmanager.net
[2021-01-24 15:10:20.014 10382M] CNAME header: Question was <IN> <AAAA> settings-win.data.microsoft.com
[2021-01-24 15:10:20.016 10382M] **** query 50026 is duplicate of 50024
``` 

I *guess* they were all blocked?

**Comment by DL6ER** [comment]
Actually, just by looking at your logs and without having reproduced this locally (yet), I *think* FTL is correct and this duplication check in `dnsmasq` may allow some queries to evade blocking in this very scenario. Nice catch. Solving this will be a bit more work but we'll have it done until the next release.

I'm a bit curious why sending one and the same query repeated times withing milliseconds (fast than the upstream response!) happens *at all* for you. I'm not seeing this at all from any client in my home network. Can you narrow this down to particular clients and, if so, operating systems? Or do they all do it for you?

**Comment by r2ixuz** [comment]
What is now the difference between "Retried" and "already forwarded"  status? Is this when the queries incoming via IPv4 and IPv6 from the same client ?

> I'm a bit curious why sending one and the same query repeated times withing milliseconds (fast than the upstream response!) happens _at all_ for you. I'm not seeing this at all from any client in my home network. Can you narrow this down to particular clients and, if so, operating systems? Or do they all do it for you?

From my testing I can say, that Windows 10 since some major versions does send each DNS query twice to the same upstream if only one DNS server IP address is set in the windows network settings. But these were actually shown as "Retried" status in PiHole.

**Comment by DL6ER** [comment]
> What is now the difference between "Retried" and "already forwarded" status?

I agree, the difference is subtle:

- `retried` means that one-and-the-same query was retried by one-and-the-same client. This means that the same query (with the same internal query ID, set by the client) is sent multiple times.
- `already forwarded` means that a new query to the same domain has been requested before the answer arrived from upstream. This does not need to be from the same client and also doesn't need to have the same ID.

I hope this makes it clear. Retried can be seen as a very special case of `already forwarded`. Both shouldn't happen in health networks, however, retried certainly means something is wrong (either the client is too impatient or the upstream too slow).

So what Windows does here is sending the same query (with the same query ID) multiple times. This is clearly a bug as there can never be a situation where this brings any advantage (except, of course, you assume a 50% loss in your local network...). Did you configure you one DNS server or did you enter the same address in two fields, I wonder? I'm not actually familiar with how the Windows network settings looks like.

**Comment by DL6ER** [comment]
Okay, some good news - nothing is evading blocking. I just verified this using Wireshark:
![Screenshot at 2021-01-24 16-34-54](https://user-images.githubusercontent.com/16748619/105635202-1f4edc80-5e62-11eb-8866-0f459ef64a76.png)
The recorded data has the following appearance in the Query Log so exactly what @Daxtorim saw:
![Screenshot at 2021-01-24 16-36-13](https://user-images.githubusercontent.com/16748619/105635247-4ad1c700-5e62-11eb-9726-d65d518cb959.png)

So yes, everything is still blocked.

**Comment by DL6ER** [comment]
(Sorry for the many posts in series, but I want to ensure you see this - and you wouldn't get informed about edits)

Fixed pushed. The same test queries now lead to:
![Screenshot at 2021-01-24 16-42-35](https://user-images.githubusercontent.com/16748619/105635520-81f4a800-5e63-11eb-993e-67ac0ac02945.png)

It was indeed a simple displaying issue and hence an easy fix.

**Comment by Daxtorim** [comment]
I see `already forwarded` from all my clients (windows, android, iOS), except the newest iOS devices, but they haven't been super active today, so not much data on that. Windows is the worst offender, but it's also the most active client at any time of the day. 

My Fritz!Box only allows one local DNS server for each IPv4 and IPv6 to be set in the DHCP settings. After skimming through the log files it looks like all `already forwarded` queries were send from the clients IPv4 *and* IPv6 address each. Windows does that a lot, although not all the time. And as said above, I'm seeing the same behavior from Android and iOS, although much less often.

**Comment by r2ixuz** [comment]
> Did you configure you one DNS server or did you enter the same address in two fields, I wonder? I'm not actually familiar with how the Windows network settings looks like.

My DHCP is offering only PiHole as DNS server IP address and only IPv4. (FritzBox Router can only deliver one DNS server via DHCP...) So then Windows has no secondary DNS configured which leads to these "Retried" status. To avoid this, you can disable Windows' own DNS caching service via Registry but then some things might not work. Better is to set DNS manually and the secondary DNS server IP to an unused local IP address to workaround this misbehaviour.

**Comment by DL6ER** [comment]
Maybe you can report this bug, maybe they will solve it... Until they (may or not) happen we should be prepared for this now. Do you see anything else which is odd with the branches in here?

I do have a Fritzbox and some Android devices (neither Windows nor Apple) myself and am not observing this. *However*, I do not use the Fritzbox as my DHCP server because of a defect in the mesh system when your Mesh Master (7590, living room, DECT base) is not your Internet router (7530, basement, DSL). I had to chose this particular setup because if the 7530 would be the Mesh Master while the 7590 would be the DECT base, I would not be able to use the telephones properly. And using the 7530 as Mesh Master + DECT base isn't an option either as then I have no DECT signal upper floor. The Fritzbox Mesh ecosystem really doesn't like if the Master is not also the DHCP server - and the DHCP server *has to be* the DSL-router (or, well, anohter device altogether). So I cannot use their DHCP server and this seems to save me from such defects in my local network. But that just as an aside.

**Comment by r2ixuz** [comment]
> > What is now the difference between "Retried" and "already forwarded" status?
> 
> I agree, the difference is subtle:
> 
> * `retried` means that one-and-the-same query was retried by one-and-the-same client. This means that the same query (with the same internal query ID, set by the client) is sent multiple times.
> * `already forwarded` means that a new query to the same domain has been requested before the answer arrived from upstream. This does not need to be from the same client and also doesn't need to have the same ID.

Thanks for clarification!

> I hope this makes it clear. Retried can be seen as a very special case of `already forwarded`. Both shouldn't happen in health networks, however, retried certainly means something is wrong (either the client is too impatient or the upstream too slow).

I think speaking about that this would normally only occur on unhealthy networks is not entirely true. Sometimes you have DNS requests which takes a long time for the upstream servers to resolve. But there are DNS client implentations outside which do only wait for a couple of 100 ms for a response. For example open http://taiwan.cn in a Chrome browser. Pihole needs over 500ms to get an answer from DoT Google DNS which leads to these "already forward" status because Chrome already send the request again. I guess it is not a problem but not completely theoretical.

**Comment by DL6ER** [comment]
> I think speaking about that this would normally only occur on unhealthy networks is not entirely true. Sometimes you have DNS requests which takes a long time for the upstream servers to resolve.

Well, I'd not say that such long delays are typical. I see that requesting the page you asked for takes roughly 3.5 seconds for me running my own recursive resolver. I still think that this out of the ordinary. You typically visit pages from either your own continent or one that is really good connected. Anything beyond 200 msec is abnormal, more so if you are using a third-party DNS resolver which may/should have this already in cache. But, yes, I agree, my standpoint should be weakened in this special case.

**Comment by Daxtorim** [comment]
@DL6ER 
>Do you see anything else which is odd with the branches in here?

Pi-hole version is v5.2.4 (Latest: v5.2.4)
AdminLTE version is update/dnsmasq-v2.84 v5.3.1-2-g2d02855 (Latest: v5.3.1)
FTL version is update/dnsmasq-v2.84 vDev-cff76ba (Latest: v5.5.1)

![grafik](https://user-images.githubusercontent.com/72616153/105644924-b6cf2200-5e98-11eb-9e99-aa57d49083b6.png)

Now the duplicate queries are given status "2" instead of the new "14", but the upstream server is missing. It no longer says `already forwarded` but instead says `OK (forwarded to )`. No idea what Windows was thinking when it issued six (!) queries for api.github.com in less than 150 ms.

---

I don't think there is any new info in the logs but for completeness' sake:

#### /var/log/pihole.log
```
Jan 24 23:02:59 dnsmasq[28259]: query[A] api.github.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 23:02:59 dnsmasq[28259]: forwarded api.github.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:600)
Jan 24 23:02:59 dnsmasq[28259]: query[AAAA] api.github.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 23:02:59 dnsmasq[28259]: forwarded api.github.com to 127.0.0.1 (/root/project/src/dnsmasq/forward.c:600)
Jan 24 23:02:59 dnsmasq[28259]: query[A] api.github.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1631)
Jan 24 23:02:59 dnsmasq[28259]: query[AAAA] api.github.com from 192.168.178.20 (/root/project/src/dnsmasq/forward.c:1631)
Jan 24 23:02:59 dnsmasq[28259]: reply api.github.com is 140.82.121.5 (/root/project/src/dnsmasq/cache.c:493)
Jan 24 23:02:59 dnsmasq[28259]: reply api.github.com is NODATA-IPv6 (/root/project/src/dnsmasq/cache.c:493)
Jan 24 23:02:59 dnsmasq[28259]: query[A] api.github.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 23:02:59 dnsmasq[28259]: cached api.github.com is 140.82.121.5 (/root/project/src/dnsmasq/rfc1035.c:1771)
Jan 24 23:02:59 dnsmasq[28259]: query[AAAA] api.github.com from fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (/root/project/src/dnsmasq/forward.c:1638)
Jan 24 23:02:59 dnsmasq[28259]: cached api.github.com is NODATA-IPv6 (/root/project/src/dnsmasq/rfc1035.c:1751)
```

---
#### /var/log/pihole-FTL.log
```
[2021-01-24 23:02:59.696 28259M] **** new UDP query[A] query "api.github.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 161, FTL 94775, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 23:02:59.697 28259M] api.github.com is not known
[2021-01-24 23:02:59.703 28259M] **** forwarded api.github.com to 127.0.0.1#5335 (ID 161, /root/project/src/dnsmasq/forward.c:602)
[2021-01-24 23:02:59.703 28259M] **** new UDP query[AAAA] query "api.github.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 162, FTL 94776, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 23:02:59.704 28259M] api.github.com is not known
[2021-01-24 23:02:59.707 28259M] **** forwarded api.github.com to 127.0.0.1#5335 (ID 162, /root/project/src/dnsmasq/forward.c:602)
[2021-01-24 23:02:59.756 28259M] **** new UDP query[A] query "api.github.com" from wlan0:192.168.178.20 (ID 163, FTL 94777, /root/project/src/dnsmasq/forward.c:1633)
[2021-01-24 23:02:59.757 28259M] api.github.com is not known
[2021-01-24 23:02:59.758 28259M] **** query for api.github.com is already in progress (ID 163)
[2021-01-24 23:02:59.758 28259M] **** new UDP query[AAAA] query "api.github.com" from wlan0:192.168.178.20 (ID 164, FTL 94778, /root/project/src/dnsmasq/forward.c:1633)
[2021-01-24 23:02:59.759 28259M] api.github.com is not known
[2021-01-24 23:02:59.760 28259M] **** query for api.github.com is already in progress (ID 164)
[2021-01-24 23:02:59.783 28259M] api.github.com is known as not to be blocked
[2021-01-24 23:02:59.783 28259M] CNAME api.github.com
[2021-01-24 23:02:59.783 28259M] **** got reply api.github.com is 140.82.121.5 (ID 161, /root/project/src/dnsmasq/cache.c:494)
[2021-01-24 23:02:59.784 28259M]      Flags: F_FORWARD F_IPV4
[2021-01-24 23:02:59.787 28259M] **** query 94777 is duplicate of 94775
[2021-01-24 23:02:59.812 28259M] **** got reply api.github.com is (NODATA) (ID 162, /root/project/src/dnsmasq/cache.c:494)
[2021-01-24 23:02:59.813 28259M]      Flags: F_FORWARD F_NEG F_IPV6
[2021-01-24 23:02:59.814 28259M] **** query 94778 is duplicate of 94776
[2021-01-24 23:02:59.816 28259M] **** new UDP query[A] query "api.github.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 165, FTL 94779, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 23:02:59.817 28259M] api.github.com is known as not to be blocked
[2021-01-24 23:02:59.817 28259M] **** got cache answer for api.github.com / 140.82.121.5 / <unknown> (ID 165, /root/project/src/dnsmasq/rfc1035.c:1774)
[2021-01-24 23:02:59.817 28259M]      Flags: F_FORWARD F_IPV4
[2021-01-24 23:02:59.818 28259M] api.github.com is known as not to be blocked
[2021-01-24 23:02:59.818 28259M] CNAME api.github.com ---> api.github.com
[2021-01-24 23:02:59.821 28259M] **** new UDP query[AAAA] query "api.github.com" from wlan0:fddc:704f:99b9:178:a4bd:67b4:3a05:88f8 (ID 166, FTL 94780, /root/project/src/dnsmasq/forward.c:1640)
[2021-01-24 23:02:59.822 28259M] api.github.com is known as not to be blocked
[2021-01-24 23:02:59.822 28259M] **** got cache answer for api.github.com /  / <unknown> (ID 166, /root/project/src/dnsmasq/rfc1035.c:1752)
[2021-01-24 23:02:59.822 28259M]      Flags: F_FORWARD F_NEG F_IPV6
```

**Comment by Daxtorim** [comment]
I think I found the problem. I left a comment on the pull request: https://github.com/pi-hole/FTL/pull/1040

**Comment by pandapaddel** [comment]
Its still not working for me and it seems there was still a problem in dnsmasq:
http://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=commit;h=141a26f979b4bc959d8e866a295e24f8cf456920

**Comment by yubiuser** [comment]
@pandapaddel see
https://github.com/pi-hole/FTL/pull/1071
